samples: add sample code for ConversationProfile, Conversation and Participant

deliaqi opened the pull request:
Hi, this is Jiaqi Liu from CCAI - Assist and Knowledge team. We are trying to add java sample code for APIs related to CCAI-features, like answer record management and conversation management. This is the first PR for updating answer record, could you help with reviewing it? Let me know anything else needed for checking in the code. Thanks!

deliaqi committed:
add UpdateAnswerRecord sample code

chingor13 commented on samples/snippets/src/main/java/com/example/dialogflow/AnswerRecordManagement.java:
```diff
@@ -0,0 +1,73 @@
+      boolean clicked)
+      throws ApiException, IOException {
+    try (AnswerRecordsClient answerRecordsClient = AnswerRecordsClient.create()) {
+      String answerRecordName = 
+        "projects/" + projectId + "/locations/global/answerRecords/" + answerRecordId;
```
If there is a resource-name class available for use, we should use it rather than building resource name strings.

```suggestion
      AnswerRecordName answerRecordName = 
        AnswerRecordName.ofProjectLocationAnswerRecordName(projectId, "global", answerRecordId);
```

deliaqi replied:
The doc https://googleapis.dev/java/google-cloud-dialogflow/latest/com/google/cloud/dialogflow/v2/AnswerRecordName.html#ofProjectLocationAnswerRecordName-java.lang.String-java.lang.String-java.lang.String- says it's a BetaApi, should we still use it?

deliaqi replied:
That‘s possible. I have made this change.

chingor13 commented on samples/snippets/src/main/java/com/example/dialogflow/AnswerRecordManagement.java:
```diff
@@ -0,0 +1,73 @@
+   * @throws ApiException
+   * @throws IOException
+   */
+  /**
+   * 
+   * @param projectId
+   * @param answerRecordId
+   * @param clicked
+   * @return
+   * @throws ApiException
+   * @throws IOException
+   */
```
This looks duplicated

deliaqi replied:
Will clear it.

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/UpdateAnswerRecordTest.java:
```diff
@@ -0,0 +1,79 @@
+      for (AnswerRecord answerRecord : 
+      response.iterateAll()) {
+        if (!answerRecord.getAnswerFeedback().getClicked()) {
+          answerRecordId = answerRecord.getName().split("/")[5];
```
You should be able to use `AnswerRecordName.parse(string)`

deliaqi replied:
I tried AnswerRecordName.parse(string) but it seems to return the whole answer record name. Do you mean it should return the answer record id?

deliaqi replied:
Got it. Thanks.

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/UpdateAnswerRecordTest.java:
```diff
@@ -0,0 +1,79 @@
+  @After
+  public void tearDown() throws IOException {
+    // Reset unclicked answer record
+    AnswerRecordManagement.updateAnswerRecord(PROJECT_ID, answerRecordId, false);
```
This expects a long-lived resource to be setup on the test project? What will happen if there are two tests running simultaneously?

deliaqi replied:
They might cause some conflicts. Another way to do this is to prepare enough answerRecord with false(default) clicked value, so that we do not need to reset it. (Actually I just realized that the clicked value cannot be set back to false.) Do you know more details about the internal testing setup (like project id and existing answer records) of this repo for checking sample code PR?

chingor13 replied:
Generally, unless we need a long-lived resource, we create a new one (with a uuid-based identifier) in the setup and delete it during teardown.

This makes it so multiple tests can run in parallel and so any developer can run the test against their own project without some external setup.

deliaqi replied:
This test does expect "unclicked" answer records for given conversation profiles. Is it possible to prepare them as long-lived resource? 
Another way to do it is to create new conversation profile, conversations, participants and try to create answer record for each test, but in this case we still need preconfigured data including knowledge bases to trigger the creation of new answer record.

deliaqi replied:
The current approach is to get one answer record for each test, is it doable? The assumption is that the testing project has at least one AnswerRecord in it.

chingor13 commented on samples/snippets/src/main/java/com/example/dialogflow/AnswerRecordManagement.java:
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 Google LLC
```
```suggestion
 * Copyright 2022 Google LLC
```

New year :)

deliaqi replied:
Got it ^

chingor13 reviewed (changes requested):
I think according to the guidelines, you should add a static main method with sample values.

deliaqi committed:
update code style

deliaqi committed:
update code style

deliaqi committed:
update code style

deliaqi committed:
add location for updating answer record

deliaqi commented:
Hi Jeff, could you review the updated code again? Let me know if you have other comments!

chingor13 commented on samples/snippets/src/main/java/com/example/dialogflow/AnswerRecordManagement.java:
```diff
@@ -0,0 +1,62 @@
+import com.google.protobuf.FieldMask;
+import java.io.IOException;
+
+public class AnswerRecordManagement {
```
The [sample guidelines](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#method-structure) require having a static main method that calls the meat of the sample that can be run directly by the developer/IDE.

deliaqi replied:
Done

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/UpdateAnswerRecordTest.java:
```diff
@@ -0,0 +1,84 @@
+  private static final String LOCATION = "global";
+
+  private static void requireEnvVar(String varName) {
+    assertNotNull(String.format(varName), String.format(varName));
```
Should this be reading the env var from the system?

deliaqi replied:
Yes, sorry for the type, just updated it.

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/UpdateAnswerRecordTest.java:
```diff
@@ -0,0 +1,84 @@
+    Optional<AnswerRecord> answerRecordOptional = getOneAnswerRecord();
+    assertTrue(answerRecordOptional.isPresent());
+
+    AnswerRecord originalAnswerRecord = answerRecordOptional.get();
+    String answerRecordId = AnswerRecordName.parse(originalAnswerRecord.getName()).getAnswerRecord();
+    boolean originalClickedValue = originalAnswerRecord.getAnswerFeedback().getClicked();
+    boolean newClickedValue = !originalClickedValue;
+
+    // Update clicked value
+    AnswerRecord updatedAnswerRecord = AnswerRecordManagement.updateAnswerRecord(
+        PROJECT_ID, LOCATION, answerRecordId, newClickedValue);
+    assertEquals(newClickedValue, updatedAnswerRecord.getAnswerFeedback().getClicked());
+
+    // Reset clicked value
+    AnswerRecord resetAnswerRecord = AnswerRecordManagement.updateAnswerRecord(
+        PROJECT_ID, LOCATION, answerRecordId, originalClickedValue);
+    assertEquals(originalClickedValue, resetAnswerRecord.getAnswerFeedback().getClicked());
+    assertEquals(originalAnswerRecord, resetAnswerRecord);
```
This still has the possibility for race conditions if there are tests running at the same time.

Is it possible to create a unique record in the test setup so that simultaneous test runs are working on different resources?

deliaqi replied:
Answer Records are created when there's a human agent assistant suggestion generated, and there is no public APIs to create a unique record. One way to do it is to create a conversation and send conversation messages to Dialogflow that could trigger a new suggestion. But this would require that the testing environment has correct setup (conversation profile, knowledge base, etc.) for at least one agent assist feature, do we currently support it? 
https://cloud.google.com/dialogflow/priv/docs/reference/rest/v2/AnswerRecord
https://cloud.google.com/dialogflow/es/docs/reference/rest/v2/projects.answerRecords

deliaqi committed:
add static main method and correct requireEnvVar method

deliaqi committed:
add sample code for conversation, conversation profile and participant

deliaqi commented:
Try to add sample code that requires less long-alive resources first after discussing with @chingor13. The starting point can be sample code for ConversationProfile, Conversation and Participant.
I have updated the code and removed the previous code for AnswerRecord, could you review the PR again?

deliaqi committed:
correct requireEnvVar method and typo

deliaqi committed:
setup Before and After each test

deliaqi committed:
add Copyright descriptions

chingor13 commented on samples/snippets/src/main/java/com/example/dialogflow/ConversationProfileManagement.java:
```diff
@@ -0,0 +1,252 @@
+      SuggestionConfig.Builder suggestionConfigBuilder = SuggestionConfig.newBuilder();
+
+      // Add knowledge base for article suggestion
+      if (articleSuggestionKnowledgeBaseId.isPresent()) {
+        KnowledgeBaseName articleSuggestionKbName = 
+            KnowledgeBaseName.of(projectId, articleSuggestionKnowledgeBaseId.get());
+        SuggestionFeatureConfig articleSuggestionFeatureConfig = 
+            SuggestionFeatureConfig.newBuilder()
+              .setSuggestionFeature(
+                SuggestionFeature.newBuilder().setType(Type.ARTICLE_SUGGESTION).build())
+              .setSuggestionTriggerSettings(
+                SuggestionTriggerSettings.newBuilder()
+                  .setNoSmalltalk(true).setOnlyEndUser(true).build())
+              .setQueryConfig(
+                SuggestionQueryConfig.newBuilder()
+                  .setKnowledgeBaseQuerySource(
+                    KnowledgeBaseQuerySource.newBuilder()
+                    .addKnowledgeBases(articleSuggestionKbName.toString()))
+                  .setMaxResults(3).build())
+              .build();
+        suggestionConfigBuilder.addFeatureConfigs(articleSuggestionFeatureConfig);
+      }
```
These should _probably_ be separate similar snippets.

deliaqi replied:
I see. In this case maybe it's redundant to have list/get/delete examples for each resource, I will keep create method for now and update the structure. Would it make sense?

chingor13 reviewed (changes requested):
From https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#class-structure: `Each snippet should be be contained in its own file, within a class with a name descriptive of the snippet and a similarly named method`

deliaqi committed:
keep one snippet in each file

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationProfileTest.java:
```diff
@@ -0,0 +1,102 @@
+                .getQueryConfig().getKnowledgeBaseQuerySource().getKnowledgeBases(0)
+            ).getKnowledgeBase().equals(faqKnowledgeBaseId)));
+
+    // Delete the conversation profile
+    String conversationProfileId = ConversationProfileName.parse(
+        createdConversationProfile.getName()).getConversationProfile();
+    try (ConversationProfilesClient conversationProfilesClient = ConversationProfilesClient.create()) {
+      ConversationProfileName conversationProfileName = 
+          ConversationProfileName.ofProjectLocationConversationProfileName(
+            PROJECT_ID, location, conversationProfileId);
+      conversationProfilesClient.deleteConversationProfile(conversationProfileName.toString());
+    }
```
You might want to move this to the `tearDown`. It's possible to leak resources if the creation succeeds, but the test assertion fails and thus the deletion never runs.

You could do this by maintaining a resource name to cleanup and in the teardown only delete it if it was set.

deliaqi replied:
Good point, thanks

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/CreateParticipantTest.java:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  @Test
+  public void testUCreateParticipant() throws IOException {
```
```suggestion
  public void testCreateParticipant() throws IOException {
```

Was this intentional or a typo?

deliaqi replied:
Sorry it's a typo. Updated

deliaqi committed:
clean up conversation profile after testing

chingor13 commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationProfileTest.java:
```diff
@@ -40,21 +41,38 @@
 
   private static final String PROJECT_ID = System.getenv("GOOGLE_CLOUD_PROJECT");
+  private static final String LOCATION = "global";
+  private static String conversationProfileNameToDelete = "";
```
`null` is probably a better choice for empty than empty string

deliaqi replied:
Yes thanks! Done.

deliaqi committed:
update code style

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

deliaqi committed:
set output stream for testing

lesv commented on samples/snippets/src/main/java/com/example/dialogflow/ConversationProfileManagement.java:
```diff
@@ -0,0 +1,137 @@
+    String conversationProfileDisplayName = "my-conversation-profile-display-name";
+
+    // Create a conversation profile
+    String articleSuggestionKnowledgeBaseId = "my-fake-article-suggestion-knowledge-base-id";
+    String faqKnowledgeBaseId = "my-fake-faq-knowledge-base-id";
```
Should this suggest how a user can determine real values?

deliaqi replied:
Yes currently it's confusing, I added more comments here to describe how to create those resources.

lesv commented on samples/snippets/src/main/java/com/example/dialogflow/ConversationProfileManagement.java:
```diff
@@ -0,0 +1,137 @@
+    // TODO(developer): Replace these variables before running the sample.
+    String projectId = "my-project-id";
+    String location = "my-location";
+    String conversationProfileDisplayName = "my-conversation-profile-display-name";
```
Should this mention where and how it's used?

lesv commented on samples/snippets/src/main/java/com/example/dialogflow/ConversationProfileManagement.java:
```diff
@@ -0,0 +1,137 @@
+      Optional<String> articleSuggestionKnowledgeBaseId,
+      Optional<String> faqKnowledgeBaseId)
+      throws ApiException, IOException {
+    try (ConversationProfilesClient conversationProfilesClient =
```
whitespace before `try`?

lesv replied:
What line width are yous set to?  I think we are at least 100 for Java, and maybe 120 for devrel.

deliaqi replied:
100 for Java, it is Google's recommendation right?

lesv commented on samples/snippets/src/main/java/com/example/dialogflow/ConversationProfileManagement.java:
```diff
@@ -0,0 +1,137 @@
+      if (articleSuggestionKnowledgeBaseId.isPresent()) {
+        KnowledgeBaseName articleSuggestionKbName =
+            KnowledgeBaseName.of(projectId, articleSuggestionKnowledgeBaseId.get());
+        SuggestionFeatureConfig articleSuggestionFeatureConfig =
+            SuggestionFeatureConfig.newBuilder()
+                .setSuggestionFeature(
+                    SuggestionFeature.newBuilder().setType(Type.ARTICLE_SUGGESTION).build())
+                .setSuggestionTriggerSettings(
+                    SuggestionTriggerSettings.newBuilder()
+                        .setNoSmalltalk(true)
+                        .setOnlyEndUser(true)
+                        .build())
+                .setQueryConfig(
+                    SuggestionQueryConfig.newBuilder()
+                        .setKnowledgeBaseQuerySource(
+                            KnowledgeBaseQuerySource.newBuilder()
+                                .addKnowledgeBases(articleSuggestionKbName.toString()))
+                        .setMaxResults(3)
+                        .build())
+                .build();
```
Wow - that looks complicated.  Is there an easier way to present this since we are trying to teach users here?

deliaqi replied:
Functionality-wise, it's hard to simplify it as we should include such nested configurations to make the feature work, but we can improve the readability by adding variables one by one and also add more descriptions.

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationProfileTest.java:
```diff
@@ -0,0 +1,139 @@
+
+  @BeforeClass
+  public static void checkRequirements() {
+    requireEnvVar("GOOGLE_APPLICATION_CREDENTIALS");
```
Probably shouldn't be a requirement these days, since we specifically recommend users don't use SA's locally.

deliaqi replied:
Yes it's kind of risky, but currently the sample code suggests it, should we keep the consistency for now or refactor a new authentication method for the current testing environment?
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#testing-setup

lesv replied:
@kolea2 FYI - we should change our standard.

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationProfileTest.java:
```diff
@@ -0,0 +1,139 @@
+      conversationProfileNameToDelete = null;
+    }
+
+    System.setOut(null);
```
VERY BAD PRACTICE, please restore STDOUT.

deliaqi replied:
OK will do, thanks for the suggestions!

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationProfileTest.java:
```diff
@@ -0,0 +1,139 @@
+                        .equals(faqKnowledgeBaseId)));
+
+    // Delete the conversation profile
+    deleteConversationProfile(conversationProfileNameToDelete);
+    conversationProfileNameToDelete = null;
```
It looks like a failure will leave a lot of conversation Profile's around.  Perhaps you'd like to make things a bit more robust.

deliaqi replied:
It could be mitigated by checking `conversationProfileNameToDelete` during `tearDown`. Do you have other suggestions?

lesv replied:
We often look for >24h old resources in the `setUp()` method and delete them.  That way if we crash, they are eventually removed.

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationTest.java:
```diff
@@ -0,0 +1,98 @@
+
+  @BeforeClass
+  public static void checkRequirements() {
+    requireEnvVar("GOOGLE_APPLICATION_CREDENTIALS");
```
Maybe not required.

deliaqi replied:
It is actually required for running tests.
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.kokoro/tests/run_tests.sh#L58

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateConversationTest.java:
```diff
@@ -0,0 +1,98 @@
+      conversationProfilesClient.deleteConversationProfile(conversationProfileName.toString());
+    }
+
+    System.setOut(null);
```
VERY BAD PRACTICE, please restore to a saved value.

deliaqi replied:
Will do.

lesv commented on samples/snippets/src/test/java/com/example/dialogflow/CreateParticipantTest.java:
```diff
@@ -0,0 +1,106 @@
+      conversationProfilesClient.deleteConversationProfile(conversationProfileName.toString());
+    }
+
+    System.setOut(null);
```
VERY BAD PRACTICE

deliaqi replied:
Will do.

lesv reviewed (changes requested):
* Region Tags are missing
* Tests can be in a single file if that makes things easier.
* Never `System.setOut(null);`
* Please read [SAMPLE_FORMAT](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)

deliaqi committed:
avoid returning value for snippet methods

deliaqi commented:
The suggestion about making one single test file is reasonable, maybe we can add an integration test file, including the workflow from creating a conversation profile to creating participants,  while the current separate unit tests could still make sense for testing each component.

lesv commented:
@deliaqi Please internally visit github/ and join `googleapis`

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

lesv committed:
Merge branch 'main' into main

deliaqi commented:
> @deliaqi Please internally visit github/ and join `googleapis`

Thanks, done!

lesv commented:
Assuming owlbot runs, you should be able to merge.

deliaqi committed:
remove code about faq assist feature

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md
